File: 54. Spiral Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if (not matrix) or (not matrix[0]):
            return []
        m = len(matrix)
        n = len(matrix[0])
        ret = []
        k = 0
        while len(ret) < m*n:
            i, j = k, k
            while j < n-k:
                ret.append(matrix[i][j])
                j += 1
            
            i, j = k+1, n-k-1
            while i < m-k:
                ret.append(matrix[i][j])
                i += 1
            
            i, j = m-1-k, n-k-2
            while i>k and j >= k:
                ret.append(matrix[i][j])
                j -= 1
            
            i, j = m-1-k-1, k
            while i>k and j<n-k-1:
                ret.append(matrix[i][j])
                i -= 1
            
            k += 1
        
        return ret
            

File: 54. Spiral Matrix.go
func spiralOrder(matrix [][]int) []int {
	m, n := len(matrix), len(matrix[0])
	res := make([]int, 0, m*n)

	x0, y0, x1, y1 := 0, 0, m-1, n-1
	x, y := 0, 0

	for len(res) < cap(res) {
		if x0 == x1 && y0 == y1 {
			res = append(res, matrix[x0][y0])
			break
		}
		for y < y1 && len(res) < cap(res) {
			res = append(res, matrix[x][y])
			y++
		}
		for x < x1 && len(res) < cap(res) {
			res = append(res, matrix[x][y])
			x++
		}
		for y > y0 && len(res) < cap(res) {
			res = append(res, matrix[x][y])
			y--
		}
		for x > x0 && len(res) < cap(res) {
			res = append(res, matrix[x][y])
			x--
		}
		x0, y0, x1, y1 = x0+1, y0+1, x1-1, y1-1
		x, y = x0, y0
	}
	return res
}
